### PR TITLE
Add IPv4 IGMPv2 support

### LINKED ISSUE #2336: IPv4 - Multicast Join/Leave Support
**_Reported by Gajinder Vij:_**

Add protocol support for joining and leaving multicast groups, as specified in RFC 3376 and 4606.

(Imported from Jira ZEP-829)

### PR DESCRIPTION
**Pull request by jukkar**
Fixes #2336

### CODE REVIEW

**Comment by jukkar** [comment]
Please help reviewing this so we could get it to 2.6, thanks in advance!

File: subsys/net/ip/Kconfig.ipv4 [hunk 1]
```diff
@@ -46,6 +46,16 @@ config NET_IPV4_ACCEPT_ZERO_BROADCAST
...
 
+config NET_IPV4_IGMP
+	bool "Internet Group Management Protocol (IGMP) support"
+	select NET_IPV4_HDR_OPTIONS
+	default y
+	help
+	  The value depends on your network needs. IGMP should normally
+	  be active. Currently we support IGMPv2 and earlier versions.
+	  Requires IPv4 header support because IP Router Alert option
+	  must be sent. See RFC 2236 for details.
```
Comment (mniestroj) [inline subsys/net/ip/Kconfig.ipv4:57]: I am a bit concerned about `default y`. Is there any added value if application does not use multicast directly?

Personally I was using Zephyr with cellular modems (using PPP) and WiFi to build IoT devices. In both cases I was more or less constrained in flash size and mostly wanted to reach cloud using TCP (HTTP, MQTT) or UDP (CoAP). So I wonder if IGMP is any added value (from what I just learned, it seems to be useful only for multicast). So I would either prefer it disabled by default to not "bloat" default networking configuration too much or alternatively improve this help message a bit to encourage users to disable it in specific cases (like using only unicast connections). At the end I don't want Zephyr networking to be considered "heavy" by new (or existing) users and expecting users to know when to disable rarely used features (like IGMP?).

Comment (jukkar) [reply subsys/net/ip/Kconfig.ipv4:57]: > I am a bit concerned about `default y`. Is there any added value if application does not use multicast directly?

According to RFC 2236, "It is required to be implemented by all hosts wishing to receive IP multicasts". So we could make it optional if we consider that multicast support needs to be enabled separately. There has not been much complaints about this missing IGMP support so this does not seem to be videly used feature among current Zephyr users. So I am ok if we make this optional.

> it seems to be useful only for multicast)

Absolutely, this is only meant for IPv4 multicast.

> At the end I don't want Zephyr networking to be considered "heavy" by new (or existing) users

I agree.

**Comment by jukkar** [comment]
* Updated to latest upstream
* IGMP support is disabled by default

**Review by mniestroj — APPROVED** [review]
LGTM

**Comment by jukkar** [comment]
@all If anyone has time, help would be needed to review this before the merge window closes.

**Review by rlubos — APPROVED** [review]
LGTM

**Pull request merged**